pkg/signals/signals.go

```diff
@@ -0,0 +1,34 @@
+package signals
+
+import (
+	"os"
+	"os/signal"
+	"sync"
+)
+
+// On registers a function to be called when one of sigs is received. If no sigs are provided On
+// is a noop. It returns a cancellation func that can be used to unregister the signals. The
+// cancellation func is idempotent.
+func On(fn func(), sigs ...os.Signal) func() {
```

Comment:
I like this method. The idea of abstracting the code that gets executed on signals from how that registering is done (in Go) is beautiful.

But, when multiple functions are registered for the same signal, how do we guarantee that they all have time to run if they call `os.Exit` (something quite common for certain signals)? Usually you try to have only one shutdown handler and maybe "wait" an arbitrary amount of time before exiting for other subroutine shutdown.

Maybe we don't need to solve this problem just yet since we only have one. But I wonder if we can do something smart to orchestrate them since we are "making" them all go through this method.

Reply:
Thanks for the feedback.

> how do we guarantee that they all have time to run if they call os.Exit

We could probably have a lengthy discussion on this. For now, I've purposely made this the callers responsibility. The rationale is rooted in the fact all use-cases at present are quite simplistic and so the responsibility falls on the consumer to ensure no 1 handler calls `os.Exit()` in a way that upsets other handlers.

> maybe "wait" an arbitrary amount of time before exiting for other subroutine shutdown

That sounds like a graceful shutdown. Again, I think that's on the consumer to orchestrate, not this registration function. I made a comment to the effect of "we can do better" in `root.go` FWIW.